ql/lib/codeql/kaleidoscope/Dataflow.qll

```diff
@@ -0,0 +1,530 @@
+     * For more information, see
+     * [Locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
+     */
+    predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      this.getLocation().hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    }
+  }
+
+  class ExprNode extends Node, TExprNode {
+    private DataFlowExpr expr;
+
+    ExprNode() { this = TExprNode(expr) }
+
+    Cfg::Node getCfgNode() { result = expr }
+
+    override string toString() { result = expr.toString() }
+
+    override Location getLocation() { result = expr.getLocation() }
+  }
+
+  class ParameterNode extends Node, TParameterNode {
+    private Identifier parameter;
+
+    ParameterNode() { this = TParameterNode(parameter) }
+
+    predicate isParameterOf(DataFlowCallable c, ParameterPosition pos) {
+      parameter = c.(FunctionDefinition).getPrototype().getArgument(pos)
+    }
+
+    override string toString() { result = parameter.toString() }
+
+    override Location getLocation() { result = parameter.getLocation() }
+
+    Identifier getParameter() { result = parameter }
+  }
+
+  class ArgumentNode extends ExprNode {
+    ArgumentNode() {
+      this.getCfgNode().getAstNode() = any(FunctionCallExpression e).getArgument(_) or
+      this.getCfgNode().getAstNode() = any(BinaryOpExpression e).getLhs() or
+      this.getCfgNode().getAstNode() = any(BinaryOpExpression e).getRhs() or
+      this.getCfgNode().getAstNode() = any(UnaryOpExpression e).getOperand()
+    }
+
+    predicate argumentOf(DataFlowCall call, ArgumentPosition pos) {
+      this.getCfgNode() = call.(Cfg::Node).getAPredecessor+() and
+      (
+        call.(Cfg::Node).getAstNode() =
+          any(FunctionCallExpression e | e.getArgument(pos) = this.getCfgNode().getAstNode()) or
+        call.(Cfg::Node).getAstNode() =
+          any(BinaryOpExpression e |
+            pos = 0 and e.getLhs() = this.getCfgNode().getAstNode()
+            or
+            pos = 1 and e.getRhs() = this.getCfgNode().getAstNode()
+          ) or
+        call.(Cfg::Node).getAstNode() =
+          any(UnaryOpExpression e | pos = 0 and e.getOperand() = this.getCfgNode().getAstNode())
+      )
+    }
+  }
+
+  class ReturnNode extends Node, TReturningNode {
+    private Cfg::Node node;
+
+    ReturnNode() { this = TReturningNode(node) }
+
+    ReturnKind getKind() { result = TNormalReturn() }
+
+    override string toString() { result = "return " + node.toString() }
+
+    override Location getLocation() { result = node.getLocation() }
+  }
+
+  newtype TReturnKind = TNormalReturn()
+
+  class ReturnKind extends TReturnKind {
+    /** Gets a textual representation of this element. */
+    string toString() { result = "return" }
+  }
+
+  /**
+   */
+  class SsaDefinitionNode extends Node, TSsaDefinitionNode {
+    Ssa::Definition def;
```

Comment:
Why not use `Ext` nodes like in Ruby?

Reply:
I used the "swift" implementation as a guide line. The Ext variants looks slightly more complicated. Is it worth switching? Also why does the Ssa library define two sets of classes, if one of them is supposedly better?